sysmonforlinux.c

```diff
@@ -543,6 +543,78 @@ void processProcessAccess(CONST PSYSMON_EVENT_HEADER eventHdr)
     DispatchEvent(newEventHdr);
 }
 
+//--------------------------------------------------------------------
```

Comment:
We need a similar method for FileDelete that also handles hashing (processFileDelete).

Reply:
Actually, FileDelete presents a different challenge. Let's focus on process create for now.

Reply:
@eeriedusk  When testing out your latest changes, I get a SIGBUS error when using a config that enables file deletion events:

<Sysmon schemaversion="4.81">
        <EventFiltering>
                <ProcessCreate onmatch="exclude" />
                <ProcessTerminate onmatch="exclude" />
                <FileCreate onmatch="exclude" />
                <FileDelete onmatch="exclude" />
        </EventFiltering>
</Sysmon>

The stacktrace of the SIGBUS is:

1573                **HashType = SysmonCryptoCurrent();         // Default is sha256
(gdb) bt
#0  LinuxGetHashTypeInformation(SYSMON_EVENT_TYPE_FMT*, SYSMON_EVENT_HEADER*, unsigned int**, unsigned int*)
    (EventType=0x5565fc652320 <SYSMONEVENT_FILE_DELETE_Type>, EventHeader=0x7f62ffee300c, HashType=0x7ffe495612a0, FileIndex=0x7ffe49561290)
    at /home/m/hashfeature/SysmonForLinux/sysmonCommon/eventsCommon.cpp:1573
#1  0x00005565fc5eaa5f in EventResolveField(PLARGE_INTEGER, PSYSMON_EVENT_TYPE_FMT, PSYSMON_DATA_DESCRIPTOR, PSYSMON_EVENT_HEADER, ULONG, PEVENT_DATA_DESCRIPTOR, BOOLEAN)
    (Time=0x7f62ffee3034, EventType=0x5565fc652320 <SYSMONEVENT_FILE_DELETE_Type>, EventBuffer=0x7ffe495716d0, EventHeader=0x7f62ffee300c, FieldIndex=7, Output=0x7ffe49571480, ForceOutputString=true)
    at /home/m/hashfeature/SysmonForLinux/sysmonCommon/eventsCommon.cpp:1732
#2  0x00005565fc5ec06e in EventProcess(SYSMON_EVENT_TYPE_FMT*, SYSMON_DATA_DESCRIPTOR*, SYSMON_EVENT_HEADER*, unsigned long*)
    (EventType=0x5565fc652320 <SYSMONEVENT_FILE_DELETE_Type>, EventBuffer=0x7ffe495716d0, EventData=0x7f62ffee300c, UserSid=0x0) at /home/m/hashfeature/SysmonForLinux/sysmonCommon/eventsCommon.cpp:2441
#3  0x00005565fc5ed042 in DispatchEvent(PVOID) (event=0x7f62ffee300c) at /home/m/hashfeature/SysmonForLinux/sysmonCommon/eventsCommon.cpp:2995
#4  0x00005565fc5b626b in handleEvent (ctx=0x0, cpu=11, data=0x7f62ffee300c, size=188) at /home/m/hashfeature/SysmonForLinux/sysmonforlinux.c:657
#5  0x00007f6332ec32f1 in perf_buffer__process_record (e=<optimized out>, ctx=<optimized out>) at libbpf.c:11925
#6  0x00007f6332ec3442 in perf_event_read_simple
    (mmap_mem=0x7f62ffee2000, mmap_size=67108864, page_size=<optimized out>, copy_mem=0x5565feed9070, copy_size=0x5565feed9078, private_data=0x5565feed9060, fn=0x7f6332ec32a0 <perf_buffer__process_record>)
    at libbpf.c:11553
#7  0x00007f6332ed1b2b in perf_buffer__process_records (pb=<optimized out>, pb=0x5565feee5870, cpu_buf=<optimized out>) at libbpf.c:11947
#8  perf_buffer__poll (pb=0x5565feee5870, timeout_ms=timeout_ms@entry=1000) at libbpf.c:11972
#9  0x00007f6332eb62ea in telemetryStart
    (ebpfConfig=0x7ffe49572b10, eventCb=0x5565fc5b6137 <handleEvent>, eventsLostCb=0x5565fc5b6281 <handleLostEvents>, telemetryReady=<optimized out>, telemetryReloadConfig=0x5565fc5b6d94 <configChange>, context=0x0, argv=0x7ffe495735e8, fds=0x5565fd264f30 <mapFds>) at /home/m/pmc/SysinternalsEBPF/telemetryLoader.c:1526
#10 0x00005565fc5b8525 in main (argc=4, argv=0x7ffe495735e8) at /home/m/hashfeature/SysmonForLinux/sysmonforlinux.c:1562

The SIGBUS is as a result of a file deletion event that tries to set the hash type:

 **HashType = SysmonCryptoCurrent();         // Default is sha256

Since we are unable to modify the original event from the eBPF perf buffer we should first make a copy of the event and set the hash type in the copy (similar to the processProcessCreate function you added). I would recommend adding a processFileDelete function and call it from handleEvent. We'll still not be able to get the file hash of a file deletion due to the file being deleted by the time we get the event in user mode (separate problem) but we won't fault with a SIGBUS.

Reply:
Indeed, I figured out the same. About the FileDelete event crash, I added the processFileDelete call in handleEvent on my side, I'll push it soon. I also noticed it would be possible to race condition the handleEvent calls. In our case, it's possible to replace the logged file before hitting handleEvent in order to mislead the hash function and result in a wrong file hash. However, it's not an exact science and requires a bit of luck. Therefore, I don't know if this kind of issue is part of your scope...

Reply:
Something to look into for sure. For now (I don't want to hold up your awesome work on getting file hashing for Process Create) let's just make sure that for file deletion events don't result in any hash (outputs '-') and then we can merge this PR. The sequence of events would be: 

1. I merge the SysmonCommon PR
2. You will have to update the SysmonCommon commit in Sysmon to use latest. 
3. Merge Sysmon PR

I've already updated the backend build systems to add OpenSSL so things should build fine for the PR at that point.

Reply:
To be sure FileDelete does not output any hash we would have to change:
```c
    if(OPT_SET( HashAlgorithms )){
        newEvent->m_HashType = *((unsigned int *)OPT_VALUE( HashAlgorithms ));
    }
    else{
        newEvent->m_HashType = 0;
    }
```
to a hardcoded:
```c
    newEvent->m_HashType = 0;
```
in `processFileDelete`.

Reply:
I tested the latest and it looks great! The build is failing as a result of static code analysis with the following errors: 

`
SysmonForLinux/sysmonforlinux.c:584:32: warning: Dereference of null pointer
        newEvent->m_HashType = *((unsigned int *)OPT_VALUE( HashAlgorithms ));
                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

/azp/fix/SysmonForLinux/test/linuxRules.cpp:244:23: warning: Null pointer passed to 1st parameter expecting 'nonnull'
                ptr = strchr(ptr, ' ');
                      ^~~~~~~~~~~~~~~~
/azp/fix/SysmonForLinux/test/linuxRules.cpp:410:23: warning: Null pointer passed to 1st parameter expecting 'nonnull'
                ptr = strchr(ptr, ' ');
                      ^~~~~~~~~~~~~~~~
/azp/fix/SysmonForLinux/test/linuxRules.cpp:666:16: warning: Dereference of null pointer (loaded from variable 'ptr')
        while (*ptr == ' ') {
               ^~~~
/azp/fix/SysmonForLinux/test/linuxRules.cpp:797:24: warning: Access to field 'm_EventType' results in a dereference of a null pointer (loaded from variable 'event')
    event->m_EventType = ProcessCreate;
    ~~~~~              ^
`

The way to try this is to build the same way we do in templates/build.yaml:

`
    scan-build cmake ..
    scan-build -v --status-bugs -o verifier_result --exclude ../test -disable-checker cplusplus.NewDelete,cplusplus.NewDeleteLeaks make
`

There is caveat that the static analysis may find different issues depending on the version of the tool chain being used. We use a lowest common denominator clang version 10. If you're not seeing the above errors its likely due to difference in compiler versions. Here is the dockerfile we use for our build system if that makes it easier (all pre-reqs installed):

`
FROM ubuntu:20.04

# To make it easier for build and release pipelines to run apt-get,
# configure apt to not require confirmation (assume the -y argument by default)
ENV DEBIAN_FRONTEND=noninteractive

RUN apt-get update && apt-get install -y gnupg wget

#install .NET and T4 text transform
RUN wget https://packages.microsoft.com/config/ubuntu/20.04/packages-microsoft-prod.deb -O packages-microsoft-prod.deb
RUN dpkg -i packages-microsoft-prod.deb
RUN rm packages-microsoft-prod.deb
RUN apt update
RUN apt -y install dotnet-sdk-6.0
RUN dotnet tool install --global dotnet-t4 --version 2.3.1


RUN apt upgrade -y \
    && apt-get install -y --no-install-recommends \
        curl \
        git \
        build-essential \
        gcc \
        g++ \
        make \
        cmake \
        libelf-dev \
        llvm \
        clang \
        libxml2 \
        libxml2-dev \
        libzstd1 \
        libgtest-dev \
        libc6-dev-i386 \
        apt-transport-https \
        dirmngr \
        googletest \
        google-mock \
        libgmock-dev \
        libjson-glib-dev \
        liblocale-gettext-perl \
        ca-certificates \
        fakeroot \
        lsb-release \
        software-properties-common \
        gettext \
        pax \
        clang-tools \
        libssl-dev

WORKDIR /azp
RUN wget https://raw.githubusercontent.com/torvalds/linux/master/include/uapi/linux/openat2.h -O /usr/include/linux/openat2.h

# install debbuild
RUN wget https://github.com/debbuild/debbuild/releases/download/22.02.1/debbuild_22.02.1-0ubuntu20.04_all.deb \
    && dpkg -i debbuild_22.02.1-0ubuntu20.04_all.deb

ENV PATH="$PATH:/root/.dotnet/tools"

SHELL ["/bin/bash"]
`